apis/server/v1alpha2/device_types.go

```diff
@@ -135,6 +150,9 @@ type DeviceParameters struct {
 	// +optional
 	UserData *string `json:"userdata,omitempty"`
 
+	// +optional
+	UserDataRef *DataKeySelector `json:"userdataRef,omitempty"`
```

Comment:
I'd suggest having a way to differentiate between the kinds, i.e. `Secret` and `ConfigMap`, this reference can point to because names are unique at kind level, so there might be unexpected results if there is `ConfigMap` and `Secret` with same name and namespace.

For references, we usually go for `FieldNameKindRef`, like `UserDataSecretRef` and/or `UserDataConfigMapRef`.

Another way could be to go for discriminator field, like `kind` field in `DataKeySelector`:
```golang
// DataKeySelector defines required spec to access a key of a configmap or secret
type DataKeySelector struct {
	NamespacedName `json:",inline,omitempty"`
	// +kubebuilder:validation:Enum=Secret;ConfigMap
	Kind           string `json:"kind"`
	Key            string `json:"key"`
	Optional       bool   `json:"optional,omitempty"`
}
```